tower-hedge/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+extern crate futures;
+extern crate hdrhistogram;
+#[macro_use]
+extern crate log;
+extern crate tokio_timer;
+extern crate tower_filter;
+extern crate tower_service;
+
+use futures::future;
+use futures::future::FutureResult;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tower_filter::Filter;
+
+mod delay;
+mod latency;
+mod rotating_histogram;
+mod select;
+
+use delay::Delay;
+use latency::Latency;
+use rotating_histogram::RotatingHistogram;
+use select::Select;
+
+type Histo = Arc<Mutex<RotatingHistogram>>;
+pub type Service<InnerSvc, P> = select::Select<
+    SelectPolicy<P>,
+    Latency<Histo, InnerSvc>,
+    Delay<DelayPolicy, Filter<Latency<Histo, InnerSvc>, PolicyPredicate<P>>>,
+>;
+type Error = Box<dyn std::error::Error + Send + Sync>;
+
+/// A policy which describes which requests can be cloned and then whether those
+/// requests should be retried.
+pub trait Policy<Request> {
+    /// clone_request is called when the request is first received to determine
+    /// if the request is retryable.
+    fn clone_request(&self, req: &Request) -> Option<Request>;
+    /// can_retry is called after the hedge timeout to determine if the hedge
+    /// retry should be issued.
+    fn can_retry(&self, req: &Request) -> bool;
+}
+
+#[derive(Clone)]
+pub struct PolicyPredicate<P>(P);
+pub struct DelayPolicy {
+    histo: Histo,
+    latency_percentile: f32,
+}
+pub struct SelectPolicy<P> {
+    policy: P,
+    histo: Histo,
+    min_data_points: u64,
+}
+
+/// A middleware that pre-emptively retries requests which have been outstanding
+/// for longer than a given latency percentile.  If either of the original
+/// future or the retry future completes, that value is used.
+pub fn service<S, P, Request>(
+    service: S,
```

Comment:
This is more of a personal preference: since there are lot of arguments here, what do you think of a `Builder` style instead? For instance:

```rust
let svc = tower::hedge::service(svc, policy, 10, 0.9, Duration::from_secs(10));

// vs

let svc = tower::hedge::Builder::new(svc, policy)
    .min_data_points(10)
    .latency_percentile(0.9)
    .period(Duration::from_secs(10))
    .build();
```

Reply:
Builders feel a bit boilerplate-y to me and 5-6 arguments isn't *that* many...